File: shipNavigation/frontend/src/geo.js
// Geodesic math on the WGS84 ellipsoid — pure JS port of backend/app/geo.py.
// No dependencies; mirrors the backend so route length and the drawn line
// always use the same formulas.

export const A = 6378137.0
export const F = 1.0 / 298.257223563
export const B = A * (1.0 - F)
export const NM_TO_M = 1852.0
export const MEAN_R = 6371009.0
const DEG = Math.PI / 180.0

export function normLon(lon) { return ((lon + 180) % 360 + 360) % 360 - 180 }

const clamp = (v, lo, hi) => Math.min(hi, Math.max(lo, v))

function fringe(deg) { return ((deg % 360) + 360) % 360 }

// Return { az1, az2, distanceM } (azimuths in degrees, 0..360).
export function vincentyInverse(lat1, lon1, lat2, lon2) {
  const phi1 = lat1 * DEG
  const phi2 = lat2 * DEG
  const lam0 = normLon(lon2 - lon1) * DEG

  const u1 = Math.atan((1.0 - F) * Math.tan(phi1))
  const u2 = Math.atan((1.0 - F) * Math.tan(phi2))
  const sinU1 = Math.sin(u1); const cosU1 = Math.cos(u1)
  const sinU2 = Math.sin(u2); const cosU2 = Math.cos(u2)

  let lam = lam0
  let lamPrev = lam
  let sinSigma = 0; let cosSigma = 0; let sigma = 0; let sinAlpha = 0
  let cosSqAlpha = 0; let cos2SigmaM = 0; let deltaSigma = 0
  let converged = false

  for (let it = 0; it < 200; it += 1) {
    const sinLam = Math.sin(lam)
    const cosLam = Math.cos(lam)
    sinSigma = Math.sqrt((cosU2 * sinLam) ** 2 + (cosU1 * sinU2 - sinU1 * cosU2 * cosLam) ** 2)
    if (sinSigma < 1e-14) return { az1: 0, az2: 0, distanceM: 0 }
    cosSigma = sinU1 * sinU2 + cosU1 * cosU2 * cosLam
    sigma = Math.atan2(sinSigma, cosSigma)
    sinAlpha = (cosU1 * cosU2 * sinLam) / sinSigma
    cosSqAlpha = 1.0 - sinAlpha * sinAlpha
    cos2SigmaM = cosSqAlpha ? cosSigma - (2.0 * sinU1 * sinU2) / cosSqAlpha : 0.0
    const cc = (F / 16.0) * cosSqAlpha * (4.0 + F * (4.0 - 3.0 * cosSqAlpha))
    lamPrev = lam
    lam = lam0 + (1.0 - cc) * F * sinAlpha * (
      sigma + cc * sinSigma * (cos2SigmaM + cc * cosSigma * (-1.0 + 2.0 * cos2SigmaM ** 2))
    )
    if (Math.abs(lam - lamPrev) < 1e-12) { converged = true; break }
  }

  if (!converged) {
    const s = haversine(lat1, lon1, lat2, lon2)
    const az1 = Math.atan2(Math.sin(lam) * cosU2, cosU1 * sinU2 - sinU1 * cosU2 * Math.cos(lam)) / DEG
    const az2 = Math.atan2(Math.sin(lam) * cosU1, -sinU1 * cosU2 + cosU1 * sinU2 * Math.cos(lam)) / DEG
    return { az1: fringe(az1), az2: fringe(az2), distanceM: s }
  }

  const uSq = cosSqAlpha * (A * A - B * B) / (B * B)
  const aa = 1.0 + (uSq / 16384.0) * (4096.0 + uSq * (-768.0 + uSq * (320.0 - 175.0 * uSq)))
  const bb = (uSq / 1024.0) * (256.0 + uSq * (-128.0 + uSq * (74.0 - 47.0 * uSq)))
  deltaSigma = bb * sinSigma * (
    cos2SigmaM
    + (bb / 4.0) * (
      cosSigma * (-1.0 + 2.0 * cos2SigmaM ** 2)
      - (bb / 6.0) * cos2SigmaM * (-3.0 + 4.0 * sinSigma ** 2) * (-3.0 + 4.0 * cos2SigmaM ** 2)
    )
  )
  const s = B * aa * (sigma - deltaSigma)

  const az1 = Math.atan2(cosU2 * Math.sin(lamPrev), cosU1 * sinU2 - sinU1 * cosU2 * Math.cos(lamPrev)) / DEG
  const az2 = Math.atan2(cosU1 * Math.sin(lamPrev), -sinU1 * cosU2 + cosU1 * sinU2 * Math.cos(lamPrev)) / DEG
  return { az1: fringe(az1), az2: fringe(az2), distanceM: s }
}

// Return [lat2, lon2] from start + initial azimuth + distance.
export function vincentyDirect(lat1, lon1, az1Deg, distM) {
  if (distM <= 0.0) return [lat1, lon1]
  const phi1 = lat1 * DEG
  const lam1 = lon1 * DEG
  const alpha1 = az1Deg * DEG

  const sinA1 = Math.sin(alpha1); const cosA1 = Math.cos(alpha1)
  const tanU1 = (1.0 - F) * Math.tan(phi1)
  const cosU1 = 1.0 / Math.sqrt(1.0 + tanU1 * tanU1)
  const sinU1 = tanU1 * cosU1
  const sigma1 = Math.atan2(tanU1, cosA1)
  const sinAlpha = cosU1 * sinA1
  const cosSqAlpha = 1.0 - sinAlpha * sinAlpha
  const uSq = cosSqAlpha * (A * A - B * B) / (B * B)
  const aa = 1.0 + (uSq / 16384.0) * (4096.0 + uSq * (-768.0 + uSq * (320.0 - 175.0 * uSq)))
  const bb = (uSq / 1024.0) * (256.0 + uSq * (-128.0 + uSq * (74.0 - 47.0 * uSq)))

  let sigma = distM / (B * aa)
  let cos2SigmaM = 0.0
  for (let it = 0; it < 200; it += 1) {
    cos2SigmaM = Math.cos(2.0 * sigma1 + sigma)
    const ss = Math.sin(sigma); const cs = Math.cos(sigma)
    const ds = bb * ss * (
      cos2SigmaM
      + (bb / 4.0) * (
        cs * (-1.0 + 2.0 * cos2SigmaM ** 2)
        - (bb / 6.0) * cos2SigmaM * (-3.0 + 4.0 * ss ** 2) * (-3.0 + 4.0 * cos2SigmaM ** 2)
      )
    )
    const sigmaNew = distM / (B * aa) + ds
    if (Math.abs(sigmaNew - sigma) < 1e-12) { sigma = sigmaNew; break }
    sigma = sigmaNew
  }

  const ss = Math.sin(sigma); const cs = Math.cos(sigma)
  const xx = sinU1 * ss - cosU1 * cs * cosA1
  const phi2 = Math.atan2(
    sinU1 * cs + cosU1 * ss * cosA1,
    (1.0 - F) * Math.sqrt(sinAlpha * sinAlpha + xx * xx),
  )
  const lam = Math.atan2(ss * sinA1, cosU1 * cs - sinU1 * ss * cosA1)
  const cc = (F / 16.0) * cosSqAlpha * (4.0 + F * (4.0 - 3.0 * cosSqAlpha))
  const ll = lam - (1.0 - cc) * F * sinAlpha * (
    sigma
    + cc * ss * (cos2SigmaM + cc * cs * (-1.0 + 2.0 * cos2SigmaM ** 2))
  )
  return [phi2 / DEG, normLon((lam1 + ll) / DEG)]
}

export function haversine(lat1, lon1, lat2, lon2) {
  const p1 = lat1 * DEG; const p2 = lat2 * DEG
  const dp = p2 - p1
  const dl = normLon(lon2 - lon1) * DEG
  const h = Math.sin(dp / 2) ** 2 + Math.cos(p1) * Math.cos(p2) * Math.sin(dl / 2) ** 2
  return 2.0 * MEAN_R * Math.asin(Math.sqrt(clamp(h, 0, 1)))
}

export function geodesicDistanceNm(lat1, lon1, lat2, lon2) {
  return vincentyInverse(lat1, lon1, lat2, lon2).distanceM / NM_TO_M
}

// Dense polyline along the geodesic between two points, spaced ~stepNm apart
// (default 30nm). Endpoints are exact; the last one is exactly [lat2, lon2].
export function sampleGeodesic(lat1, lon1, lat2, lon2, stepNm = 30) {
  const { az1, distanceM } = vincentyInverse(lat1, lon1, lat2, lon2)
  if (distanceM <= 1.0) return [[lat1, lon1]]
  const stepM = stepNm * NM_TO_M
  const n = Math.max(1, Math.ceil(distanceM / stepM))
  const pts = [[lat1, lon1]]
  for (let i = 1; i < n; i += 1) {
    const d = distanceM * i / n
    pts.push(vincentyDirect(lat1, lon1, az1, d))
  }
  pts.push([lat2, lon2])
  return pts
}

// Length of a polyline in nautical miles, measured with the same geodesic math
// that generates the line — so the number matches what is drawn.
export function polylineLengthNm(points) {
  let total = 0.0
  for (let i = 1; i < points.length; i += 1) {
    const [la1, lo1] = points[i - 1]
    const [la2, lo2] = points[i]
    total += geodesicDistanceNm(la1, lo1, la2, lo2)
  }
  return total
}

// Convert a [[lat, lon], ...] polyline into continuous longitudes (can exceed
// +/-180) so a great circle that crosses the antimeridian stays unbroken.
// Renders correctly after splitIntoRenderedParts().
export function unwrapLongitudes(points) {
  const out = [[points[0][0], points[0][1]]]
  let carry = 0
  for (let i = 1; i < points.length; i += 1) {
    let lon = points[i][1] + carry
    let d = lon - out[i - 1][1]
    while (d > 180) { carry -= 360; lon -= 360; d = lon - out[i - 1][1] }
    while (d < -180) { carry += 360; lon += 360; d = lon - out[i - 1][1] }
    out.push([points[i][0], lon])
  }
  return out
}

// Split continuous-longitude points into renderable LineString parts, each in
// the [-180, 180] world copy, breaking exactly where the route crosses the
// antimeridian. Returns [[[lon, lat], ...], ...] (Note: lon-lat order).
export function splitIntoRenderedParts(points) {
  const parts = []
  let part = []
  let shift = 0
  for (const [lat, lon] of points) {
    let lr = lon - shift
    let nextShift = shift
    if (lr > 180) nextShift += 360
    else if (lr <= -180) nextShift -= 360
    if (nextShift !== shift) {
      if (part.length) parts.push(part)
      part = []
      shift = nextShift
      lr = lon - shift
    }
    part.push([lr, lat])
  }
  if (part.length) parts.push(part)
  return parts
}

File: shipNavigation/backend/app/geo.py
"""Pure-python geodesic math on the WGS84 ellipsoid.

- vincenty_inverse : distance + initial/final azimuths between two points
- vincenty_direct  : propagate a point by a distance along an azimuth
- sample_geodesic  : dense polyline between two points (used by Stage A)
- haversine        : fast spherical distance (used for A* heuristics)

No external deps here so the routing math is self-contained.
"""

from __future__ import annotations

import math

A = 6378137.0
F = 1.0 / 298.257223563
B = A * (1.0 - F)
NM_TO_M = 1852.0


def _norm_lon(lon: float) -> float:
    return ((lon + 180.0) % 360.0) - 180.0


def vincenty_inverse(lat1: float, lon1: float, lat2: float, lon2: float):
    """Return (initial_azimuth_deg, final_azimuth_deg, distance_m)."""
    phi1 = math.radians(lat1)
    phi2 = math.radians(lat2)
    lam = math.radians(_norm_lon(lon2 - lon1))

    u1 = math.atan((1.0 - F) * math.tan(phi1))
    u2 = math.atan((1.0 - F) * math.tan(phi2))
    sin_u1, cos_u1 = math.sin(u1), math.cos(u1)
    sin_u2, cos_u2 = math.sin(u2), math.cos(u2)

    lam_prev = lam
    sin_sigma = cos_sigma = sigma = sin_alpha = cos_sq_alpha = cos_2_sigma_m = delta_sigma = 0.0
    converged = False
    for _ in range(200):
        sin_lam = math.sin(lam)
        cos_lam = math.cos(lam)
        sin_sigma = math.sqrt(
            (cos_u2 * sin_lam) ** 2 + (cos_u1 * sin_u2 - sin_u1 * cos_u2 * cos_lam) ** 2
        )
        if sin_sigma < 1e-14:
            return 0.0, 0.0, 0.0
        cos_sigma = sin_u1 * sin_u2 + cos_u1 * cos_u2 * cos_lam
        sigma = math.atan2(sin_sigma, cos_sigma)
        sin_alpha = (cos_u1 * cos_u2 * sin_lam) / sin_sigma
        cos_sq_alpha = 1.0 - sin_alpha * sin_alpha
        cos_2_sigma_m = cos_sigma - 2.0 * sin_u1 * sin_u2 / cos_sq_alpha if cos_sq_alpha else 0.0
        cc = (F / 16.0) * cos_sq_alpha * (4.0 + F * (4.0 - 3.0 * cos_sq_alpha))
        lam_prev = lam
        lam = math.radians(_norm_lon(lon2 - lon1)) + (1.0 - cc) * F * sin_alpha * (
            sigma
            + cc
            * sin_sigma
            * (cos_2_sigma_m + cc * cos_sigma * (-1.0 + 2.0 * cos_2_sigma_m**2))
        )
        if abs(lam - lam_prev) < 1e-12:
            converged = True
            break

    if not converged:
        # Antipodal / degenerate: fall back to haversine on a sphere.
        s = haversine(lat1, lon1, lat2, lon2)
        az1 = math.degrees(math.atan2(
            math.sin(lam) * cos_u2,
            cos_u1 * sin_u2 - sin_u1 * cos_u2 * math.cos(lam),
        ))
        az2 = math.degrees(math.atan2(
            math.sin(lam) * cos_u1,
            -sin_u1 * cos_u2 + cos_u1 * sin_u2 * math.cos(lam),
        ))
        return az1 % 360.0, az2 % 360.0, s

    u_sq = cos_sq_alpha * (A * A - B * B) / (B * B)
    aa = 1.0 + (u_sq / 16384.0) * (4096.0 + u_sq * (-768.0 + u_sq * (320.0 - 175.0 * u_sq)))
    bb = (u_sq / 1024.0) * (256.0 + u_sq * (-128.0 + u_sq * (74.0 - 47.0 * u_sq)))
    delta_sigma = bb * sin_sigma * (
        cos_2_sigma_m
        + (bb / 4.0)
        * (
            cos_sigma * (-1.0 + 2.0 * cos_2_sigma_m**2)
            - (bb / 6.0) * cos_2_sigma_m * (-3.0 + 4.0 * sin_sigma**2) * (-3.0 + 4.0 * cos_2_sigma_m**2)
        )
    )
    s = B * aa * (sigma - delta_sigma)

    az1 = math.degrees(math.atan2(
        cos_u2 * math.sin(lam_prev), cos_u1 * sin_u2 - sin_u1 * cos_u2 * math.cos(lam_prev)
    ))
    az2 = math.degrees(math.atan2(
        cos_u1 * math.sin(lam_prev), -sin_u1 * cos_u2 + cos_u1 * sin_u2 * math.cos(lam_prev)
    ))
    return az1 % 360.0, az2 % 360.0, s


def vincenty_direct(lat1: float, lon1: float, az1: float, dist_m: float) -> tuple[float, float]:
    """Given start + azimuth + distance, return (lat2, lon2)."""
    if dist_m <= 0.0:
        return lat1, lon1
    phi1 = math.radians(lat1)
    lam1 = math.radians(lon1)
    alpha1 = math.radians(az1)

    sin_a1, cos_a1 = math.sin(alpha1), math.cos(alpha1)
    tan_u1 = (1.0 - F) * math.tan(phi1)
    cos_u1 = 1.0 / math.sqrt(1.0 + tan_u1 * tan_u1)
    sin_u1 = tan_u1 * cos_u1
    sigma1 = math.atan2(tan_u1, cos_a1)
    sin_alpha = cos_u1 * sin_a1
    cos_sq_alpha = 1.0 - sin_alpha * sin_alpha
    u_sq = cos_sq_alpha * (A * A - B * B) / (B * B)
    aa = 1.0 + (u_sq / 16384.0) * (4096.0 + u_sq * (-768.0 + u_sq * (320.0 - 175.0 * u_sq)))
    bb = (u_sq / 1024.0) * (256.0 + u_sq * (-128.0 + u_sq * (74.0 - 47.0 * u_sq)))

    sigma = dist_m / (B * aa)
    sin_sigma_cos = cos_2_sigma_m = 0.0
    for _ in range(200):
        cos_2_sigma_m = math.cos(2.0 * sigma1 + sigma)
        sin_sigma, cos_sigma = math.sin(sigma), math.cos(sigma)
        delta_sigma = bb * sin_sigma * (
            cos_2_sigma_m
            + (bb / 4.0)
            * (
                cos_sigma * (-1.0 + 2.0 * cos_2_sigma_m**2)
                - (bb / 6.0) * cos_2_sigma_m * (-3.0 + 4.0 * sin_sigma**2) * (-3.0 + 4.0 * cos_2_sigma_m**2)
            )
        )
        sigma_new = dist_m / (B * aa) + delta_sigma
        if abs(sigma_new - sigma) < 1e-12:
            sigma = sigma_new
            break
        sigma = sigma_new

    sin_sigma, cos_sigma = math.sin(sigma), math.cos(sigma)
    xx = sin_u1 * sin_sigma - cos_u1 * cos_sigma * cos_a1
    phi2 = math.atan2(
        sin_u1 * cos_sigma + cos_u1 * sin_sigma * cos_a1,
        (1.0 - F) * math.sqrt(sin_alpha * sin_alpha + xx * xx),
    )
    lam = math.atan2(sin_sigma * sin_a1, cos_u1 * cos_sigma - sin_u1 * sin_sigma * cos_a1)
    cc = (F / 16.0) * cos_sq_alpha * (4.0 + F * (4.0 - 3.0 * cos_sq_alpha))
    ll = lam - (1.0 - cc) * F * sin_alpha * (
        sigma
        + cc
        * sin_sigma
        * (cos_2_sigma_m + cc * cos_sigma * (-1.0 + 2.0 * cos_2_sigma_m**2))
    )
    lon2 = math.degrees(lam1 + ll)
    return math.degrees(phi2), _norm_lon(lon2)


def haversine(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Great-circle distance in metres on a mean-radius sphere."""
    r = 6371009.0
    p1 = math.radians(lat1)
    p2 = math.radians(lat2)
    dp = p2 - p1
    dl = math.radians(_norm_lon(lon2 - lon1))
    h = math.sin(dp / 2) ** 2 + math.cos(p1) * math.cos(p2) * math.sin(dl / 2) ** 2
    return 2.0 * r * math.asin(math.sqrt(min(1.0, h)))


def geodesic_distance_nm(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    _, _, m = vincenty_inverse(lat1, lon1, lat2, lon2)
    return m / NM_TO_M


def sample_geodesic(lat1: float, lon1: float, lat2: float, lon2: float, step_nm: float) -> list[tuple[float, float]]:
    """Dense polyline along the geodesic between two points, spacing ~step_nm."""
    az1, _, dist_m = vincenty_inverse(lat1, lon1, lat2, lon2)
    if dist_m <= 1.0:
        return [(lat1, lon1)]
    step_m = step_nm * NM_TO_M
    n = max(1, int(math.ceil(dist_m / step_m)))
    pts = [(lat1, lon1)]
    for i in range(1, n):
        d = dist_m * i / n
        lat, lon = vincenty_direct(lat1, lon1, az1, d)
        pts.append((lat, lon))
    pts.append((lat2, lon2))
    return pts


def polyline_length_nm(pts: list[tuple[float, float]]) -> float:
    total = 0.0
    for (la1, lo1), (la2, lo2) in zip(pts, pts[1:]):
        total += geodesic_distance_nm(la1, lo1, la2, lo2)
    return total